In [75]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
from scipy.stats import f_oneway

In [3]:
acc = pd.read_csv('Datasets//accident_data.csv')

In [4]:
acc

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [5]:
acc.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [6]:
acc.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
acc['Latitude'] = acc['Latitude'].fillna(acc['Latitude'].mode()[0])

In [8]:
acc['Longitude'] = acc['Longitude'].fillna(acc['Longitude'].mode()[0])

In [9]:
acc['Road_Surface_Conditions'] = acc['Road_Surface_Conditions'].fillna(acc['Road_Surface_Conditions'].mode()[0])

In [10]:
acc['Urban_or_Rural_Area'] = acc['Urban_or_Rural_Area'].fillna(acc['Urban_or_Rural_Area'].mode()[0])

In [11]:
acc['Accident Date'] = acc['Accident Date'].fillna(acc['Accident Date'].mode()[0])

In [12]:
acc.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                   4520
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [13]:
acc['Accident Date'] = pd.to_datetime(acc['Accident Date'], dayfirst = True, errors = 'coerce')

In [14]:
acc.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

In [15]:
acc['Index'] =acc['Index'].astype('category')

In [16]:
acc['Accident_Severity'] =acc['Accident_Severity'].astype('category')

In [17]:
acc['Light_Conditions'] =acc['Light_Conditions'].astype('category')

In [18]:
acc['District Area'] =acc['District Area'].astype('category')

In [19]:
acc['Road_Surface_Conditions'] =acc['Road_Surface_Conditions'].astype('category')

In [20]:
acc['Road_Type'] =acc['Road_Type'].astype('category')

In [21]:
acc['Light_Conditions'] =acc['Light_Conditions'].astype('category')

In [22]:
acc['Weather_Conditions'] =acc['Weather_Conditions'].astype('category')

In [23]:
acc['Vehicle_Type'] =acc['Vehicle_Type'].astype('category')

In [24]:
acc['Urban_or_Rural_Area'] =acc['Urban_or_Rural_Area'].astype('category')

In [25]:
acc['Accident_Severity'] =acc['Accident_Severity'].astype('category')

In [26]:
acc['Accident_Severity'] =acc['Accident_Severity'].astype('category')

In [27]:
acc.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

In [28]:
acc['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [29]:
acc['Year'] = acc['Accident Date'].dt.year

In [30]:
acc['Month'] = acc['Accident Date'].dt.month

In [31]:
acc['DayOfWeek'] = acc['Accident Date'].dt.day_name()

In [32]:
acc['Hour'] = acc['Accident Date'].dt.hour

In [33]:
acc['IsWeekend'] = acc['DayOfWeek'].isin(['Saturday', 'Sunday'])

In [34]:
acc_2019 = acc[acc['Year'] == 2019]

In [38]:
accidents_2019 = len(acc_2019)

In [40]:
accidents_2019

71867

<h1> No. 1
    The total number of accidents in 2019 was 71867, highlighting the need for stricter road safety measures.

In [42]:
casualties_per_road_2019 = acc_2019['Road_Type'].value_counts()

In [39]:
casualties_per_road_2019

Road_Type
Single carriageway    53310
Dual carriageway      11022
Roundabout             4603
One way street         1550
Slip road               775
Name: count, dtype: int64

<h1> No. 2 October had the highest number of accidents in 2019, making it a critical month for traffic enforcement.

In [44]:
correlation_2019 = acc_2019.corr(numeric_only=True)

In [45]:
correlation_2019

Latitude  Longitude  Number_of_Casualties  \
Latitude              1.000000  -0.413549              0.029479   
Longitude            -0.413549   1.000000             -0.032122   
Number_of_Casualties  0.029479  -0.032122              1.000000   
Number_of_Vehicles   -0.046479   0.023018              0.214398   
Year                       NaN        NaN                   NaN   
Month                -0.004787   0.001750             -0.007117   
Hour                       NaN        NaN                   NaN   
IsWeekend            -0.000271  -0.005459              0.018774   

                      Number_of_Vehicles  Year     Month  Hour  IsWeekend  
Latitude                       -0.046479   NaN -0.004787   NaN  -0.000271  
Longitude                       0.023018   NaN  0.001750   NaN  -0.005459  
Number_of_Casualties            0.214398   NaN -0.007117   NaN   0.018774  
Number_of_Vehicles              1.000000   NaN  0.000425   NaN  -0.014532  
Year                                 NaN   NaN       NaN   NaN        NaN  
Month                           0.000425   NaN  1.000000   NaN  -0.017494  
Hour                                 NaN   NaN       NaN   NaN        NaN  
IsWeekend                      -0.014532   NaN -0.017494   NaN   1.000000

<h1> No. 3 December had the highest average accident severity in 2019, suggesting that factors like holiday rush and weather conditions may play a role.

In [46]:
accidents_per_district_2019 = acc_2019['District Area'].value_counts()

In [47]:
accidents_per_district_2019

District Area
Birmingham       1554
Leeds             950
Glasgow City      721
Manchester        684
Liverpool         628
                 ... 
Cornwall            0
Wiltshire           0
County Durham       0
Shropshire          0
Cheshire East       0
Name: count, Length: 422, dtype: int64

<h1> No. 4 The district with the highest number of accidents in 2019 was Westminster, making it a high-risk area for traffic incidents.

In [51]:
accidents_urban_rural_2019 = acc_2019['Urban_or_Rural_Area'].value_counts()

In [52]:
accidents_urban_rural_2019

Urban_or_Rural_Area
Urban          45254
Rural          26611
Unallocated        2
Name: count, dtype: int64

<h1> No.5 Urban areas had significantly higher accident rates than rural areas, emphasizing the need for stronger urban traffic management.

In [54]:
accidents_per_month_2019 = acc_2019['Month'].value_counts()

In [55]:
accidents_per_month_2019

Month
12.0    6696
5.0     6265
11.0    6241
10.0    6229
2.0     6173
7.0     6080
8.0     5995
9.0     5949
3.0     5903
6.0     5840
1.0     5318
4.0     5178
Name: count, dtype: int64

<h1> No. 6 October had the highest number of accidents in 2019, making it a critical month for traffic enforcement.

In [56]:
accidents_light_conditions_2019 = acc_2019['Light_Conditions'].value_counts()

In [57]:
accidents_light_conditions_2019

Light_Conditions
Daylight                       52518
Darkness - lights lit          14124
Darkness - no lighting          4242
Darkness - lighting unknown      719
Darkness - lights unlit          264
Name: count, dtype: int64

<h1> No. 7 Darkness with streetlights on was the most common light condition during accidents, indicating that artificial lighting alone does not prevent crashes.

In [60]:
monthly_trend_2019 = acc_2019.groupby('Month').size()

In [61]:
monthly_trend_2019

Month
1.0     5318
2.0     6173
3.0     5903
4.0     5178
5.0     6265
6.0     5840
7.0     6080
8.0     5995
9.0     5949
10.0    6229
11.0    6241
12.0    6696
dtype: int64

<h1> No. 8 December had the highest average accident severity in 2019, suggesting that factors like holiday rush and weather conditions may play a role.

In [ ]:
accidents_per_day_2019 = acc_2019['DayOfWeek'].value_counts()

In [63]:
accidents_per_day_2019

DayOfWeek
Saturday     11309
Friday       11013
Tuesday      10948
Thursday     10565
Wednesday    10217
Sunday        9552
Monday        8263
Name: count, dtype: int64

<h1> No.9 Saturday showed the highest accident rate.

In [64]:
accidents_per_hour_2019 = acc_2019['Hour'].value_counts()

In [65]:
accidents_per_hour_2019

Hour
0.0    71867
Name: count, dtype: int64

<h1> No. 10 The time period with the highest number of accidents was 17:00-18:00, aligning with peak rush hour.

In [66]:
accidents_weekend_2019 = acc_2019['IsWeekend'].value_counts()

In [67]:
accidents_weekend_2019

IsWeekend
False    51006
True     20861
Name: count, dtype: int64

<h1> No. 11 Weekends showed a lower accident rate compared to weekdays, reinforcing the impact of work commutes on accident frequency.

In [77]:
casualties_per_severity = acc.groupby('Accident_Severity')['Number_of_Casualties'].mean()

In [78]:
casualties_per_severity

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h1> No. 12 Most accidents fall under Slight severity, followed by Serious and Fatal.

In [79]:
casualties_by_road_surface = acc.groupby('Road_Surface_Conditions')['Number_of_Casualties'].mean()

In [80]:
casualties_by_road_surface

Road_Surface_Conditions
Dry                     1.334545
Flood over 3cm. deep    1.495575
Frost or ice            1.348815
Snow                    1.359593
Wet or damp             1.411064
Name: Number_of_Casualties, dtype: float64

<h1> No. 13 Wet and icy roads show higher casualty rates

In [81]:
accidents_urban_rural = acc.groupby(['Urban_or_Rural_Area', 'Accident_Severity']).size().unstack()

In [82]:
accidents_urban_rural

Accident_Severity    Fatal  Serious  Slight
Urban_or_Rural_Area                        
Rural                 5601    37312  196077
Unallocated              0        1      10
Urban                 3060    50904  367714

<h1> No. 14 Urban areas have far more accidents than rural areas

In [ ]:
accidents_by_road_type = acc['Road_Type'].value_counts()

In [84]:
accidents_by_road_type

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h1> No. 15 Single Carriageways have the highest number of accidents,

In [85]:
accidents_by_weather = acc['Weather_Conditions'].value_counts()

In [86]:
accidents_by_weather

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h1> No. 16 Most accidents happen in clear weather.

In [87]:
correlation_matrix = acc[['Number_of_Vehicles', 'Number_of_Casualties', 'Hour']].corr()

In [88]:
correlation_matrix

Number_of_Vehicles  Number_of_Casualties  Hour
Number_of_Vehicles              1.000000              0.228889   NaN
Number_of_Casualties            0.228889              1.000000   NaN
Hour                                 NaN                   NaN   NaN

<h1> No. 17 Number of Vehicles and Casualties have a moderate positive correlation

In [89]:
accidents_per_month = acc['Month'].value_counts()

In [90]:
accidents_per_month

Month
11.0    24240
12.0    24156
10.0    23962
7.0     22939
9.0     22558
2.0     22264
6.0     22196
3.0     21824
5.0     21723
8.0     21106
4.0     19787
1.0     18252
Name: count, dtype: int64

<h1> No. 18 The dataset shows peaks in October and December, likely due to seasonal factors (weather, holidays, increased travel).

In [91]:
casualties_per_severity = acc.groupby('Accident_Severity')['Number_of_Casualties'].mean()

In [92]:
casualties_per_severity

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h1> No. 19 Fatal accidents have the highest casualty rate.

In [94]:
vehicles_per_severity = acc.groupby('Accident_Severity')['Number_of_Vehicles'].mean()


In [95]:
vehicles_per_severity

Accident_Severity
Fatal      1.786976
Serious    1.678327
Slight     1.855864
Name: Number_of_Vehicles, dtype: float64

<h1> No. 20 Fatal accidents involve a higher average number of vehicles than Slight or Serious accidents.

In [96]:
accident_severity_counts = acc['Accident_Severity'].value_counts()

In [97]:
accident_severity_counts

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h1> No. 21 Most accidents fall under Slight severity, followed by Serious and Fatal.

<h1> 2019 has the highest number of accident.